indico/modules/rb_new/client/js/modules/bookRoom/actions.js

```diff
@@ -0,0 +1,150 @@
+ * License, or (at your option) any later version.
+ *
+ * Indico is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Indico; if not, see <http://www.gnu.org/licenses/>.
+ */
+
+import createBookingURL from 'indico-url:rooms_new.create_booking';
+import fetchTimelineDataURL from 'indico-url:rooms_new.timeline';
+import fetchSuggestionsURL from 'indico-url:rooms_new.suggestions';
+
+import {indicoAxios, handleAxiosError} from 'indico/utils/axios';
+import {ajaxAction, submitFormAction} from 'indico/utils/redux';
+import {ajax as ajaxRules} from './serializers';
+import {ajax as ajaxFilterRules} from '../../serializers/filters';
+import {preProcessParameters} from '../../util';
+
+import {searchRooms as roomListSearchRooms} from '../../actions';
+
+// Booking creation
+export const CREATE_BOOKING_REQUEST = 'bookRoom/CREATE_BOOKING_REQUEST';
+export const CREATE_BOOKING_SUCCESS = 'bookRoom/CREATE_BOOKING_SUCCESS';
+export const CREATE_BOOKING_FAILED = 'bookRoom/CREATE_BOOKING_FAILED';
+
+// Checking availability of room
+export const GET_BOOKING_AVAILABILITY_REQUEST = 'bookRoom/GET_BOOKING_AVAILABILITY_REQUEST';
+export const GET_BOOKING_AVAILABILITY_SUCCESS = 'bookRoom/GET_BOOKING_AVAILABILITY_SUCCESS';
+export const GET_BOOKING_AVAILABILITY_ERROR = 'bookRoom/GET_BOOKING_AVAILABILITY_ERROR';
+export const RESET_BOOKING_AVAILABILITY = 'bookRoom/RESET_BOOKING_AVAILABILITY';
+export const SET_BOOKING_AVAILABILITY = 'bookRoom/SET_BOOKING_AVAILABILITY';
+
+// Timeline
+export const FETCH_TIMELINE_DATA_STARTED = 'bookRoom/FETCH_TIMELINE_DATA_STARTED';
+export const FETCH_TIMELINE_DATA_FAILED = 'bookRoom/FETCH_TIMELINE_DATA_FAILED';
+export const UPDATE_TIMELINE_DATA = 'bookRoom/UPDATE_TIMELINE_DATA';
+export const TOGGLE_TIMELINE_VIEW = 'bookRoom/TOGGLE_TIMELINE_VIEW';
+
+// Suggestions
+export const FETCH_SUGGESTIONS_STARTED = 'bookRoom/FETCH_SUGGESTIONS_STARTED';
+export const FETCH_SUGGESTIONS_FAILED = 'bookRoom/FETCH_SUGGESTIONS_FAILED';
+export const UPDATE_SUGGESTIONS = 'bookRoom/UPDATE_SUGGESTIONS';
+
+
+export function createBooking(args) {
+    const params = preProcessParameters(args, ajaxRules);
+    return submitFormAction(
+        () => indicoAxios.post(createBookingURL(), params),
+        CREATE_BOOKING_REQUEST, CREATE_BOOKING_SUCCESS, CREATE_BOOKING_FAILED
+    );
+}
+
+export function resetBookingState() {
+    return {type: RESET_BOOKING_AVAILABILITY};
+}
+
+export function fetchBookingAvailability(room, filters) {
+    return ajaxAction(
+        () => _fetchTimelineData(filters, [room]),
+        GET_BOOKING_AVAILABILITY_REQUEST,
+        [GET_BOOKING_AVAILABILITY_SUCCESS, SET_BOOKING_AVAILABILITY],
```

Comment:
i'd trigger the success only after setting the data. that way loading indicators only go away when the data is actually there (and depending on how they are configured avoid a flash of empty content)

Reply:
Maybe I'm misunderstanding what you're trying to say, but how do you even get to know when the data is "already there"? By passing it on to Redux you only make sure it will eventually reach the component...

Reply:
`[SET_BOOKING_AVAILABILITY, GET_BOOKING_AVAILABILITY_SUCCESS]`

That way the first dispatch sets the data, and the second one marks the request as successful (so if a spinner is shown during loading, it goes away once the actual data is there)

Reply:
I get it. My question is: does it really make much of a difference, since what the underlying code is doing is just dispatching one after the other? Does the interface have time to react between the two actions? Anyway, I'll change that.